cpp/oneapi/dal/backend/primitives/lapack/solve_dpc.cpp

```diff
@@ -70,12 +220,39 @@ sycl::event solve_system(sycl::queue& queue,
 
     auto [nxty, xty_event] = copy<ndorder::c, Float, ylayout, alloc>(queue, xty, dependencies);
     auto [nxtx, xtx_event] = copy<ndorder::c, Float, xlayout, alloc>(queue, xtx, dependencies);
+    const std::int64_t dim_xtx = xtx.get_dimension(0);
+
+    sycl::event solution_event;
 
     opt_array<Float> dummy{};
-    auto potrf_event = potrf_factorization<uplo>(queue, nxtx, dummy, { xtx_event });
-    auto potrs_event = potrs_solution<uplo>(queue, nxtx, nxty, dummy, { potrf_event, xty_event });
+    /// Note: this is wrapped in a try-catch block in order to catch a potential exception
+    /// thrown by oneMKL when the Cholesky factorization ('potrs') fails due to the matrix
+    /// not being positive-definite. In such case, it will then fall back to spectral
+    /// decomposition-based inversion, which is able to handle such problems.
+    try {
+        sycl::event potrf_event = potrf_factorization<uplo>(queue, nxtx, dummy, { xtx_event });
+        queue.wait_and_throw();
+        const Float diag_min = diagonal_minimum(queue, nxtx.get_data(), dim_xtx, potrf_event);
+        if (diag_min <= 1e-6)
+            throw mkl::lapack::computation_error("", "", 0);
```

Comment:
That should be the last portion of comments.

1. Please provide some descriptive name for 1e-6 threshold. We are trying to reduce the number of magic constants in our code.

2. Throwing oneMKL exception here looks like a hack to call the code in catch.
It would be better just to call spectral decomposition here and in catch. Maybe re-write it a bit in order not to duplicate the code with array copying.

Reply:
Added a comment about the threshold and changed the logic of the fallback to a named section.

Reply:
Maybe I do not understand something, but why this block cannot be just a function? For the sake of simplicity.

`goto` in DPC++ code looks pretty weird. And besides the performance aspect, the use of `goto` make the code harder to understand. As the execution order and the call stack can be very hard to track.

Reply:
Changed it into a function.